gobblin-hive-registration/src/main/java/org/apache/gobblin/hive/writer/HiveMetadataWriter.java

```diff
@@ -0,0 +1,287 @@
+        if (spec.getTable().getDbName().equals(dbName) && spec.getTable().getTableName().equals(tableName)) {
+          List<String> partitionValue =
+              spec.getPartition().isPresent() ? spec.getPartition().get().getValues() : Lists.newArrayList();
+          Cache<List<String>, HiveSpec> hiveSpecCache = specMaps.computeIfAbsent(tableKey,
+              s -> CacheBuilder.newBuilder()
+                  .expireAfterAccess(state.getPropAsInt(MetadataWriter.CACHE_EXPIRING_TIME,
+                      MetadataWriter.DEFAULT_CACHE_EXPIRING_TIME), TimeUnit.HOURS)
+                  .build());
+          HiveSpec existedSpec = hiveSpecCache.getIfPresent(partitionValue);
+          if (existedSpec != null) {
+            //if existedSpec is not null, it means we already registered this partition, so check whether we need to update the table/partition
+            if (gmce.getSchemaSource() == SchemaSource.SCHEMAREGISTRY) {
+              // Only in this case, we need to update the schema to the latest schema in schemaRegistry
+              schemaUpdateHelper(spec, topicName, tableKey);
+            }
+            if ((this.hiveRegister.needToUpdateTable(existedSpec.getTable(), spec.getTable())) ||
+                (spec.getPartition().isPresent() && this.hiveRegister.needToUpdatePartition(
+                    existedSpec.getPartition().get(), spec.getPartition().get()))) {
+              registerSpec(tableKey, partitionValue, spec, hiveSpecCache);
+            }
+          } else {
+            registerSpec(tableKey, partitionValue, spec, hiveSpecCache);
+          }
+        }
+      }
+    }
+  }
+
+  private void registerSpec(String tableKey, List<String> partitionValue, HiveSpec spec,
+      Cache<List<String>, HiveSpec> hiveSpecCache) {
+    HashMap<List<String>, ListenableFuture<Void>> executionMap =
+        this.currentExecutionMap.computeIfAbsent(tableKey, s -> new HashMap<>());
+    if (executionMap.containsKey(partitionValue)) {
+      try {
+        executionMap.get(partitionValue).get(timeOutSeconds, TimeUnit.SECONDS);
+      } catch (InterruptedException | ExecutionException | TimeoutException e) {
+        log.error("Error when getting the result of registration for table" + tableKey);
+        throw new RuntimeException(e);
+      }
+    }
+    executionMap.put(partitionValue, this.hiveRegister.register(spec));
+    hiveSpecCache.put(partitionValue, spec);
+  }
+
+  private void schemaUpdateHelper(HiveSpec spec, String topicName, String tableKey) throws SchemaRegistryException {
+    String newSchemaString =
+        spec.getTable().getSerDeProps().getProp(AvroSerdeUtils.AvroTableProperties.SCHEMA_LITERAL.getPropName());
+    if (newSchemaString != null) {
+      Schema newSchema = new Schema.Parser().parse(newSchemaString);
+      String newSchemaCreationTime = AvroUtils.getSchemaCreationTime(newSchema);
+      Set<String> existedSchemaCreationTimes = schemaCreationTimeMap.computeIfAbsent(tableKey, s -> new HashSet<>());
```

Comment:
I wonder why this has to be a set instead of a single value capturing the creation time of newestSchema that was determined before. What's the advantage to keep all ever-appearing schemaCreationTime ?

Reply:
Since when schema change, we cannot make sure the gmce with latest always and gmce with old schema are coming in order, there are high chances that we see gmce with old schema and latest schema are interspersed with each other. In these case, we don't want to  query schemaRegistry every time we see a different schema creation time, so cache the creationTime we have seen. I will change Set into Cache to avoid memory accumulating

Reply:
Thanks, this also helps to understand the comment around `HiveMetadataWriter.java # 223`